src/safemode_ui.cpp

```diff
@@ -0,0 +1,741 @@
+        for( int i = 0; i < iContentHeight; i++ ) {
+            for( int j = 0; j < 79; j++ ) {
+                mvwputch( w, i, j, c_black, ' ' );
+            }
+
+            for( auto &mapLine : mapPos ) {
+                mvwputch( w, i, mapLine.second, c_ltgray, LINE_XOXO );
+            }
+        }
+
+        const bool currentPageNonEmpty = !vRules[iTab].empty();
+
+        if( iTab == CHARACTER_TAB && g->u.name == "" ) {
+            vRules[CHARACTER_TAB].clear();
+            mvwprintz( w, 8, 15, c_white, _( "Please load a character first to use this page!" ) );
+        }
+
+        draw_scrollbar( w_border, iLine, iContentHeight, vRules[iTab].size(), 5 );
+        wrefresh( w_border );
+
+        calcStartPos( iStartPos, iLine, iContentHeight, vRules[iTab].size() );
+
+        // display safemode
+        for( int i = iStartPos; i < ( int )vRules[iTab].size(); i++ ) {
+            if( i >= iStartPos &&
+                i < iStartPos + ( ( iContentHeight > ( int )vRules[iTab].size() ) ?
+                                  ( int )vRules[iTab].size() : iContentHeight ) ) {
+                nc_color cLineColor = ( vRules[iTab][i].bActive ) ? c_white : c_ltgray;
+
+                sTemp.str( "" );
+                sTemp << i + 1;
+                mvwprintz( w, i - iStartPos, 1, cLineColor, "%s", sTemp.str().c_str() );
+                mvwprintz( w, i - iStartPos, 5, cLineColor, "" );
+
+                if( iLine == i ) {
+                    wprintz( w, c_yellow, ">> " );
+                } else {
+                    wprintz( w, c_yellow, "   " );
+                }
+
+                wprintz( w, ( iLine == i && iColumn == col_Rule ) ? hilite( cLineColor ) : cLineColor,
+                         "%s", ( ( vRules[iTab][i].sRule == "" ) ? _( "<empty rule>" ) : vRules[iTab][i].sRule ).c_str()
+                       );
+
+                auto attCreature = Creature::get_creature_attitude( ( Creature::Attitude )
+                                   vRules[iTab][i].attCreature );
+                mvwprintz( w, i - iStartPos, mapPos[col_Attitude] + 2,
+                           ( iLine == i && iColumn == col_Attitude ) ? hilite( cLineColor ) : cLineColor,
+                           "%s", ( attCreature.first ).c_str()
+                         );
+
+                mvwprintz( w, i - iStartPos, mapPos[col_Dist] + 2,
+                           ( iLine == i && iColumn == col_Dist ) ? hilite( cLineColor ) : cLineColor,
+                           "%s", ( !vRules[iTab][i].bExclude ) ? to_string( vRules[iTab][i].iProxyDist ).c_str() : "---"
+                         );
+
+                mvwprintz( w, i - iStartPos, mapPos[col_IE] + 2,
+                           ( iLine == i && iColumn == col_IE ) ? hilite( cLineColor ) : cLineColor,
+                           "%s", ( ( vRules[iTab][i].bExclude ) ? rm_prefix( _( "Exclude" ) ).c_str() : rm_prefix(
+                                       _( "Include" ) ).c_str() )
+                         );
```

Comment:
This code is hard to read. The non-standard naming scheme (hungarian notation itself, hungarian notation being mixed with non-hungarian notation) doesn't help.

Could be cleaned up by:
* Extracting a lambda to cover all 3 `mvwprintz` without a lot of repetition (wouldn't be easy, but doable)
* Extracting `vRules[iTab][i]` to a variable like `current_tab`
* Extracting some of the inlined data to named variables, to show what is being displayed